UniversalAlgebra/Homomorphisms/Setoid.lagda

```diff
@@ -0,0 +1,307 @@
+open import Function               using    ( _∘_    ;  id       )
+open import Level                  using    ( Level  ;  Lift     )
+open import Relation.Binary                       using    ( IsEquivalence  )
+open import Relation.Unary                        using    ( _⊆_ )
+import Relation.Binary.PropositionalEquality as PE
+
+-- Imports from the Agda Universal Algebra Library
+open import Overture.Preliminaries     using ( ∣_∣ ; ∥_∥ ; _⁻¹)
+open import Overture.Inverses          using ( IsInjective ; IsSurjective )
+open import Overture.Inverses          using ( SurjInv )
+open import Relations.Discrete         using ( ker ; kernel )
+open import Relations.Truncation       using ( is-set ; blk-uip ; is-embedding
+                                             ; monic-is-embedding|Set )
+open import Relations.Extensionality   using ( swelldef ; block-ext|uip ; pred-ext
+                                             ; SurjInvIsRightInv ; epic-factor )
+open import Algebras.Setoid   {𝑆 = 𝑆}  using ( 𝕌[_] ; SetoidAlgebra ; _̂_ ; Lift-SetoidAlg )
+open import Congruences.Setoid {𝑆 = 𝑆} using ( _∣≈_ ; Con ; IsCongruence ; _╱_)
+\end{code}
+
+### Homomorphisms for setoid algebras
+
+\begin{code}
+
+module _ {α ρᵃ : Level} (𝑨 : SetoidAlgebra α ρᵃ)
+         {β ρᵇ : Level} (𝑩 : SetoidAlgebra β ρᵇ)
+         where
+ private
+  A = 𝕌[ 𝑨 ] -- (𝕌 = forgetful functor)
+  B = 𝕌[ 𝑩 ]
+
+ compatible-op-map : ∣ 𝑆 ∣ → (A → B) → Type _
+ compatible-op-map 𝑓 h = ∀ a → h ((𝑓 ̂ 𝑨) a) ≡ (𝑓 ̂ 𝑩) (h ∘ a)
+
+ -- The property of being a homomorphism.
+ is-homomorphism : (A → B) → Type(𝓞 ⊔ 𝓥 ⊔ α ⊔ β)
+ is-homomorphism h = ∀ 𝑓  →  compatible-op-map 𝑓 h
+
+ -- The type of homomorphisms from `𝑨` to `𝑩`.
+ hom : Type(𝓞 ⊔ 𝓥 ⊔ α ⊔ β)
+ hom = Σ (A → B) is-homomorphism
+
+open PE.≡-Reasoning
+open PE renaming (cong to ≡-cong)
+
+module _ {α ρᵃ : Level} (𝑨 : SetoidAlgebra α ρᵃ)  -- (explicit 𝑨)
+         {β ρᵇ : Level} {𝑩 : SetoidAlgebra β ρᵇ}  -- (implicit 𝑩)
+         {γ ρᶜ : Level} (𝑪 : SetoidAlgebra γ ρᶜ)  -- (explicit 𝑪)
+         where
+ private
+  A = 𝕌[ 𝑨 ]  -- carrier of domain of 𝑨
+  B = 𝕌[ 𝑩 ]
+  C = 𝕌[ 𝑪 ]
+
+ -- The composition of homomorphisms is again a homomorphism.
+ ∘-hom : hom 𝑨 𝑩  →  hom 𝑩 𝑪  →  hom 𝑨 𝑪
+ ∘-hom (g , ghom) (h , hhom) = h ∘ g , Goal where
+
+  Goal : ∀ 𝑓 a → (h ∘ g)((𝑓 ̂ 𝑨) a) ≡ (𝑓 ̂ 𝑪)(h ∘ g ∘ a)
+  Goal 𝑓 a = (h ∘ g)((𝑓 ̂ 𝑨) a)     ≡⟨ ≡-cong h ( ghom 𝑓 a ) ⟩
+             h ((𝑓 ̂ 𝑩)(g ∘ a))     ≡⟨ hhom 𝑓 ( g ∘ a ) ⟩
+             (𝑓 ̂ 𝑪)(h ∘ g ∘ a)     ∎
+
+
+ ∘-is-hom : {f : A → B}{g : B → C}
+  →         is-homomorphism 𝑨 𝑩 f → is-homomorphism 𝑩 𝑪 g
+            -------------------------------------------------
+  →         is-homomorphism 𝑨 𝑪 (g ∘ f)
+
+ ∘-is-hom {f} {g} fhom ghom = ∥ ∘-hom (f , fhom) (g , ghom) ∥
+
+module _ {α ρ : Level} where
+
+ -- the identity homs
+ 𝒾𝒹 : (𝑨 : SetoidAlgebra α ρ) → hom 𝑨 𝑨
+ 𝒾𝒹 _ = id , λ 𝑓 a → refl
+
+ open Level
+
+ -- the lift hom
```

Comment:
Personally, I'd be wanting to put all the lift-related stuff in a submodule, to make them even more of a pain to use. They should be used extremely sparingly, only when absolutely necessary.